### ISSUE TITLE
Block dropout does not work on GPU

### ISSUE THREAD

**Issue by FilippoC** [clab/dynet#1451]
Hi,

using block dropout on GPU results in a segmentation fault:
- dynet: master branch (i.e. last commit is Sat Sep 15 23:30:30 2018 +0100)
- GPU: NVIDIA Titan X (Pascal)
- Cuda: 8.0

There is no specific information about GPU in the implementation, so I don't know if it fails because the implementation is not cuda-compatible or because `this->has_cuda_implemented = false;` is missing in the constructor.

Best,
Caio

Example:
```
#include "dynet/dynet.h"
#include "dynet/expr.h"
#include "dynet/training.h"
#include "dynet/param-init.h"

int main(int argc, char** argv)
{
    dynet::initialize(argc, argv);
    const unsigned size = 10u;

    dynet::ParameterCollection pc;
    auto p_W = pc.add_parameters({size, size});
    auto p_b = pc.add_parameters({size}, dynet::ParameterInitConst(0.f));

    for (unsigned i = 0 ; i < 100 ; ++i)
    {
        dynet::ComputationGraph cg;
        cg.set_immediate_compute(true);
        //cg.set_check_validity(true);
        auto W = dynet::parameter(cg, p_W);
        auto b = dynet::parameter(cg, p_b);
        auto r = dynet::random_normal(cg, {size});
        auto mlp = W * r + b;
        auto block_mlp = dynet::block_dropout(mlp, 0.5f);
        auto loss = dynet::sum_elems(block_mlp);
        cg.forward(loss);
    }

}
```

Compilation and execution on CPU (OK) and GPU (not OK):
```
[ccorro@logicgpu2 ~]$ g++ block_dropout.cpp -Wall -g -std=c++11 -ldynet
[ccorro@logicgpu2 ~]$ ./a.out --dynet-devices CPU
[dynet] random seed: 1299449813
[dynet] allocating memory: 512MB
[dynet] memory allocation done.
[ccorro@logicgpu2 ~]$ ./a.out --dynet-devices CPU,GPU:0
[dynet] initializing CUDA
[dynet] CUDA driver/runtime versions are 8.0/8.0
[dynet] Request for 1 specific GPU ...
[dynet] Device Number: 0
[dynet]   Device name: TITAN X (Pascal)
[dynet]   Memory Clock Rate (KHz): 5005000
[dynet]   Memory Bus Width (bits): 384
[dynet]   Peak Memory Bandwidth (GB/s): 480.48
[dynet]   Memory Free (GB): 12.6169/12.7816
[dynet]
[dynet] Device(s) selected: 0
[dynet] random seed: 4087215548
[dynet] allocating memory: 512MB
[dynet] memory allocation done.
Segmentation fault
```

GDB output:
```
[ccorro@logicgpu2 ~]$ gdb --args ./a.out --dynet-devices CPU,GPU:0
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-94.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /home/ccorro/a.out...done.
(gdb) r
Starting program: /home/ccorro/./a.out --dynet-devices CPU,GPU:0
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[dynet] initializing CUDA
[New Thread 0x7fffe1538700 (LWP 29149)]
[dynet] CUDA driver/runtime versions are 8.0/8.0
[dynet] Request for 1 specific GPU ...
[dynet] Device Number: 0
[dynet]   Device name: TITAN X (Pascal)
[dynet]   Memory Clock Rate (KHz): 5005000
[dynet]   Memory Bus Width (bits): 384
[dynet]   Peak Memory Bandwidth (GB/s): 480.48
[New Thread 0x7fffe0cb6700 (LWP 29164)]
[New Thread 0x7fffdbfff700 (LWP 29241)]
[dynet]   Memory Free (GB): 12.6169/12.7816
[dynet]
[Thread 0x7fffdbfff700 (LWP 29241) exited]
[Thread 0x7fffe0cb6700 (LWP 29164) exited]
[dynet] Device(s) selected: 0[New Thread 0x7fffe0cb6700 (LWP 29432)]
[New Thread 0x7fffdbfff700 (LWP 29513)]

[dynet] random seed: 3713333507
[dynet] allocating memory: 512MB
[dynet] memory allocation done.

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff55db548 in void dynet::BlockDropout::forward_dev_impl<dynet::Device_GPU>(dynet::Device_GPU const&, std::vector<dynet::Tensor const*, std::allocator<dynet::Tensor const*> > const&, dynet::Tensor&) const ()
   from /datastore/caio/repos/dynet/build/dynet/libdynet.so
Missing separate debuginfos, use: debuginfo-install glibc-2.17-157.el7_3.1.x86_64 libgcc-4.8.5-11.el7.x86_64 libstdc++-4.8.5-11.el7.x86_64
(gdb) bt
#0  0x00007ffff55db548 in void dynet::BlockDropout::forward_dev_impl<dynet::Device_GPU>(dynet::Device_GPU const&, std::vector<dynet::Tensor const*, std::allocator<dynet::Tensor const*> > const&, dynet::Tensor&) const ()
   from /datastore/caio/repos/dynet/build/dynet/libdynet.so
#1  0x00007ffff53d1ff3 in dynet::BlockDropout::forward_impl (this=0x1954b2a0, xs=std::vector of length 1, capacity 16 = {...}, fx=...) at /datastore/caio/repos/dynet/dynet/nodes-dropout.cc:186
#2  0x00007ffff515fd19 in dynet::Node::forward (this=this@entry=0x1954b2a0, xs=std::vector of length 1, capacity 16 = {...}, fx=...) at /datastore/caio/repos/dynet/dynet/dynet.cc:34
#3  0x00007ffff5176845 in dynet::SimpleExecutionEngine::incremental_forward (this=<optimized out>, i=<optimized out>) at /datastore/caio/repos/dynet/dynet/exec.cc:150
#4  0x00007ffff5162974 in dynet::ComputationGraph::set_dim_for_new_node (this=this@entry=0x7fffffffdc70, i=@0x7fffffffd9c0: 5) at /datastore/caio/repos/dynet/dynet/dynet.cc:385
#5  0x00007ffff5162b7e in dynet::ComputationGraph::add_function_node (this=this@entry=0x7fffffffdc70, node=<optimized out>, node@entry=0x1954b2a0) at /datastore/caio/repos/dynet/dynet/dynet.cc:171
#6  0x00007ffff5184b65 in add_function<dynet::BlockDropout, float&> (arguments=..., this=0x7fffffffdc70) at /datastore/caio/repos/dynet/dynet/dynet.h:780
#7  dynet::block_dropout (x=..., p=0.5) at /datastore/caio/repos/dynet/dynet/expr.cc:134
#8  0x0000000000401eba in main (argc=1, argv=0x7fffffffdfe8) at block_dropout.cpp:26
(gdb) frame 1
#1  0x00007ffff53d1ff3 in dynet::BlockDropout::forward_impl (this=0x1954b2a0, xs=std::vector of length 1, capacity 16 = {...}, fx=...) at /datastore/caio/repos/dynet/dynet/nodes-dropout.cc:186
186     DYNET_NODE_INST_DEV_IMPL(BlockDropout)
```

**Comment by pmichel31415** [comment]
Hi @FilippoC thanks for flagging, will take a look
